### PR TITLE
feat: Accept "printer" instance via opts to customise printing

### PR DESCRIPTION
**Pull request by christianscott**
I want to print the tree as markdown rather than as plaintext. With this change, I would achieve that by doing something like this:

```ts
import * as indexifier from 'indexifier';

class MarkdownPrinter extends indexifier.AbstractPrinter {
  constructor(private readonly baseUrl: string) {}
  printNode(node: indexifier.Node): string {
    return `[${node.name}](./${this.baseUrl}/${node.name})`;
  }
}

indexifier(dir, { printer: new MarkdownPrinter(baseUrl) });
```

I also considered changing the `isHtml` to `renderMode`. This ended up not being workable due to wanting to pass in `baseUrl`. It doesn't really make sense for this library to expose a `baseUrl` option so I opted for exposing `printer` instead.

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    };
+  
+    declare function indexifier(dir: string, opts?: IndexifierOpts): string;
+  
+    declare class AbstractPrinter {
+        print(node: Node): string;
+        abstract printNode(node: Node): string;
+    }
+  
+    interface Indexifier {
```
Comment (christianscott) [inline index.d.ts:30]:
```suggestion
    interface Indexifier extends typeof indexifier {
```
maybe?

Comment (joscha) [reply index.d.ts:30]: I am not actually sure - isn't that what usually the namespaces are used for? To collapse the two definitions into one? If you have this running in a typescript project with the current types that is good enough for now I think?

**Review by joscha — APPROVED** [review]

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+declare module 'indexifier' {
```
[the remark this answers is not in the document]
Comment (christianscott) [reply index.d.ts:1]: 🙈  don't tempt me

[the remark this answers is not in the document]
Comment (joscha) [reply index.d.ts:1]: what if I say pretty please?

[the remark this answers is not in the document]
Comment (joscha) [reply index.d.ts:1]: pretty please?

**Pull request merged**